hail/python/hail/matrixtable.py

```diff
@@ -539,6 +539,138 @@ class MatrixTable(ExprContainer):
+        ----------------------------------------
+        Column fields:
+            'col_idx': int32
+            'bar': int32
+        ----------------------------------------
+        Row fields:
+            'row_idx': int32
+            'foo': int32
+        ----------------------------------------
+        Entry fields:
+            'baz': int32
+        ----------------------------------------
+        Column key: None
+        Row key: ['row_idx']
+        ----------------------------------------
+        >>> mt.row.show()
+        +---------+-------+
+        | row_idx |   foo |
+        +---------+-------+
+        |   int32 | int32 |
+        +---------+-------+
+        |       0 |     1 |
+        |       1 |     2 |
+        +---------+-------+
+        >>> mt.col.show()
+        +---------+-------+
+        | col_idx |   bar |
+        +---------+-------+
+        |   int32 | int32 |
+        +---------+-------+
+        |       0 |     3 |
+        |       1 |     4 |
+        +---------+-------+
+        >>> mt.entry.show()
+        +---------+-------------+-------------+
+        | row_idx | <col 0>.baz | <col 1>.baz |
+        +---------+-------------+-------------+
+        |   int32 |       int32 |       int32 |
+        +---------+-------------+-------------+
+        |       0 |           1 |           2 |
+        |       1 |           3 |           4 |
+        +---------+-------------+-------------+
+
+        Parameters
+        ----------
+        globals : :class:`dict` from :class:`str` to :obj:`any`
+            Global fields by name
+
+        rows: :class:`dict` from :class:`str` to :class:`list` of :obj:`any`
+            Row fields by name
+
+        cols: :class:`dict` from :class:`str` to :class:`list` of :obj:`any`
+            Column fields by name
+
+        entries: :class:`dict` from :class:`str` to :class:`list` of :class:`list` of :obj:`any`
+            Matrix entries by name in the form `entry[row_idx][col_idx]`
+
+        Returns
+        -------
+        :class:`.MatrixTable`
+            A MatrixTable assembled from inputs keyed by `row_idx`
+
+        Notes
+        -----
+        - The number of rows and columns specified in `rows` and `cols` for all
+          fields must match the matrix dimensions in `entries`.
+        - This method is intended for educational and testing purposes only.
+        """
+        # General idea: build a `Table` representation matching that returned by
+        # `MatrixTable.localize_entries` and then call `_unlocalize_entries`.
+        def invert(kvs: Dict[str, Iterable[Any]]) -> List[Dict[str, Any]]:
+            return [dict(zip(kvs, vs)) for vs in zip(*kvs.values())]
+
+        def anyval(kvs):
+            return next(iter(kvs.values()))
+
+        # In the case rows or cols aren't specified, we need to infer the
+        # matrix dimensions from *an* entry. Which one isn't important as we
+        # won't enforce congruence among input dimensions.
```

Comment:
Ah, I commented below already, but I disagree, we should enforce a consistent number of rows and columns. A consistent design principle of the Hail API is that we try to avoid foot guns that silently ignore bad states.

Moreover, I think the code which verifies that is so obvious in function that it wouldn't need a comment anyway. Something like:
```
n_rows = None
n_cols = None

if rows:
    c = Counter(len(x) for x in rows.values())
    if len(c) > 1:
        raise ValueError(...)
    n_rows = next(iter(c.values()))
if cols:
    # ...
if entries:
    if n_rows and len(rows) != n_rows:
         raise ...
    # check/infer n_cols for/from entries
```